Advanced for loop tracking expressions support and bug fixes

crisbeto opened the pull request:
These changes add support for advanced tracking expressions inside `for` loop blocks, as well as a bug fix that I found along the way. Overview of the commits:

### refactor(compiler): computed for loop variables exposed on the wrong scope
Fixes that the computed for loop variables (e.g. $first and $last) were exposed on the parent scope instead of the for loop scope.

### refactor(core): add instruction to reference component instance
Adds an instruction that allows us to access the containing component instance directly instead of having to traverse the context tree. This will be necessary for the tracking function of `for` loop blocks.

### refactor(compiler): allow constants pool to track shared functions
Expands the constant pool to be able to reuse function declarations that have identical bodies.

### refactor(compiler): add support for advanced tracking expressions
These changes build on top of https://github.com/angular/angular/pull/51514 to add support for advanced expressions inside the `track` parameter of `for` loop blocks. There are two different outputs that the compiler can generate:

1. If the tracking function only references the item or `$index`, the compiler generates a pure arrow function as a constant references in the `repeaterCreate` instruction.
2. If the tracking function has references to properties outside of the `for` loop block, the compiler will rewrite those references to go through `this` and generate a function declaration. The runtime will `bind` the declaration to the current component instance so that the rewritten `this` references are resolved correctly.

Advanced tracking expression come with the following limitations to ensure the best possible performance:
1. They can only reference the item, `$index` and properties directly on the component instance. This means that there'll be an error when accessing this like local template variables and references. While we could get this to work, we would have to traverse the context tree at runtime which will degrade the performance of the loop, because it's a linear time operation that is performed on each comparison. Furthermore, allowing local references would require us re-evaluate the list when any one of them has changed.
2. Pipes aren't allowed inside the tracking function.
3. Object literals and pipes used inside the tracking expression will be recreated on each invocation.

crisbeto committed:
refactor(compiler): computed for loop variables exposed on the wrong scope

Fixes that the computed for loop variables (e.g. $first and $last) were exposed on the parent scope instead of the for loop scope.

crisbeto committed:
refactor(core): add instruction to reference component instance

Adds an instruction that allows us to access the containing component instance directly instead of having to traverse the context tree. This will be necessary for the tracking function of `for` loop blocks.

crisbeto committed:
refactor(compiler): allow constants pool to track shared functions

Expands the constant pool to be able to reuse function declarations that have identical bodies.

crisbeto commented on packages/core/src/render3/instructions/control_flow.ts:
```diff
@@ -105,15 +105,29 @@ class RepeaterMetadata {
     index: number, templateFn: ComponentTemplate<unknown>, decls: number, vars: number,
-    trackByFn: TrackByFunction<unknown>, emptyTemplateFn?: ComponentTemplate<unknown>,
-    emptyDecls?: number, emptyVars?: number): void {
+    trackByFn: TrackByFunction<unknown>, trackByUsesComponentInstance?: boolean,
```
Not a huge fan of the `trackByUsesComponentInstance` name, but I didn't want a param called `isPure`, because I expect most tracking functions to be pure and to omit the parameter if they are. I'm open to different names.

AndrewKushnir replied:
As a possible option can we add an extra function call to the generated code to bind trackby fn to a component instance? For example:

```
ɵɵrepeaterCreate(0, MyApp_For_1_Template, 1, 1, ɵɵbindToComponentInstance(trackFn));
```

In this case, the `ɵɵrepeaterCreate` would just work with a trackby function without the need to do conditional binding.

crisbeto replied:
This is something we were discussing Pawel and Alex, but we decided against it because it's more bytes.

AndrewKushnir replied:
I'd imagine that in minified form it'd be comparable:
```
// generated:
ɵɵrepeaterCreate(0, MyApp_For_1_Template, 1, 1, ɵɵbindToComponentInstance(trackFn));

// minified:
f(0,t,1,1,b(t));
```
vs:
```
// generated:
ɵɵrepeaterCreate(0, MyApp_For_1_Template, 1, 1, trackFn, true);

// minified:
f(0,t,1,1,t,!0);
```
The delta is 3 (`,!0`) vs 4 (`b()`) extra chars. Also if there is an `{:empty}` block present, an extra argument would always be an overhead, which would not be the case if we have a single track by fn argument.

crisbeto replied:
No worries, I think it's important that we discuss it. I'm starting to forget the details, but I think there was also some implications for the new signal instructions (might be misremembering). If we go for something like this, we could even do .bind(componentContext()) and avoid the instruction altogether.

AndrewKushnir replied:
Yeah, there might be multiple options that we can explore. In this case I'd vote for moving forward with what is currently implemented in the PR (so we don't block the PR on this) and perform some followup changes if needed.

AndrewKushnir commented on packages/core/src/render3/instructions/control_flow.ts:
```diff
@@ -105,15 +105,29 @@ class RepeaterMetadata {
   const hasEmptyBlock = emptyTemplateFn !== undefined;
   const hostLView = getLView();
-  const metadata = new RepeaterMetadata(hasEmptyBlock, new DefaultIterableDiffer(trackByFn));
+  const boundTrackBy = trackByUsesComponentInstance ?
+      trackByFn.bind(hostLView[DECLARATION_COMPONENT_VIEW][CONTEXT]) :
+      trackByFn;
```
nit: I think it'd be great to add a comment here that explains why we do not just bind in all cases (i.e. that we generate an arrow fn for pure fn cases).

AndrewKushnir commented on packages/core/src/render3/instructions/control_flow.ts:
```diff
@@ -105,15 +105,29 @@ class RepeaterMetadata {
+ *  component instance. If it doesn't, we can avoid rebinding it.
+ * @param emptyTemplateFn Reference to the template function of the empty block.
+ * @param emptyDecls The number of nodes, local refs, and pipes for the empty block.
+ * @param vars The number of bindings for the empty block.
```
`vars` -> `emptyVars`

AndrewKushnir reviewed (approved):
@crisbeto great work on improving trackBy 👍

crisbeto committed:
refactor(compiler): add support for advanced tracking expressions

These changes build on top of #51514 to add support for advanced expressions inside the `track` parameter of `for` loop blocks. There are two different outputs that the compiler can generate:

1. If the tracking function only references the item or `$index`, the compiler generates a pure arrow function as a constant references in the `repeaterCreate` instruction.
2. If the tracking function has references to properties outside of the `for` loop block, the compiler will rewrite those references to go through `this` and generate a function declaration. The runtime will `bind` the declaration to the current component instance so that the rewritten `this` references are resolved correctly.

Advanced tracking expression come with the following limitations to ensure the best possible performance:
1. They can only reference the item, `$index` and properties directly on the component instance. This means that there'll be an error when accessing this like local template variables and references. While we could get this to work, we would have to traverse the context tree at runtime which will degrade the performance of the loop, because it's a linear time operation that is performed on each comparison. Furthermore, allowing local references would require us re-evaluate the list when any one of them has changed.
2. Pipes aren't allowed inside the tracking function.
3. Object literals and pipes used inside the tracking expression will be recreated on each invocation.

jessicajaniuk commented:
This PR was merged into the repository by commit 05a16b973de47b4d023fb8977be7dfbefdd5ad0f.
